File: apimsdc/RootMsdc.cpp
/**
  * \file RootMsdc.cpp
  * API code for job RootMsdc (implementation)
  * \author Alexander Wirthmueller
  * \date created: 18 Dec 2018
  * \date modified: 18 Dec 2018
  */

#ifdef _WIN32
	#include "stdafx.h"
#endif

#include "RootMsdc.h"

/******************************************************************************
 class RootMsdc::DpchAppLogin
 ******************************************************************************/

RootMsdc::DpchAppLogin::DpchAppLogin(
			const string& scrJref
			, const string& username
			, const string& password
			, const bool m2mNotReg
			, const bool chksuspsess
			, const set<uint>& mask
		) :
			DpchAppMsdc(VecMsdcVDpch::DPCHAPPROOTMSDCLOGIN, scrJref)
		{
	if (find(mask, ALL)) this->mask = {SCRJREF, USERNAME, PASSWORD, M2MNOTREG, CHKSUSPSESS};
	else this->mask = mask;

	this->username = username;
	this->password = password;
	this->m2mNotReg = m2mNotReg;
	this->chksuspsess = chksuspsess;
};

string RootMsdc::DpchAppLogin::getSrefsMask() {
	vector<string> ss;
	string srefs;

	if (has(SCRJREF)) ss.push_back("scrJref");
	if (has(USERNAME)) ss.push_back("username");
	if (has(PASSWORD)) ss.push_back("password");
	if (has(M2MNOTREG)) ss.push_back("m2mNotReg");
	if (has(CHKSUSPSESS)) ss.push_back("chksuspsess");

	StrMod::vectorToString(ss, srefs);

	return(srefs);
};

void RootMsdc::DpchAppLogin::writeXML(
			xmlTextWriter* wr
		) {
	xmlTextWriterStartElement(wr, BAD_CAST "DpchAppRootMsdcLogin");
	xmlTextWriterWriteAttribute(wr, BAD_CAST "xmlns", BAD_CAST "http://www.mpsitech.com/msdc");
		if (has(SCRJREF)) writeString(wr, "scrJref", scrJref);
		if (has(USERNAME)) writeString(wr, "username", username);
		if (has(PASSWORD)) writeString(wr, "password", password);
		if (has(M2MNOTREG)) writeBool(wr, "m2mNotReg", m2mNotReg);
		if (has(CHKSUSPSESS)) writeBool(wr, "chksuspsess", chksuspsess);
	xmlTextWriterEndElement(wr);
};

/******************************************************************************
 class RootMsdc::DpchEngData
 ******************************************************************************/

RootMsdc::DpchEngData::DpchEngData() :
			DpchEngMsdc(VecMsdcVDpch::DPCHENGROOTMSDCDATA)
		{
	feedFEnsSps.tag = "FeedFEnsSps";
};

string RootMsdc::DpchEngData::getSrefsMask() {
	vector<string> ss;
	string srefs;

	if (has(SCRJREF)) ss.push_back("scrJref");
	if (has(FEEDFENSSPS)) ss.push_back("feedFEnsSps");

	StrMod::vectorToString(ss, srefs);

	return(srefs);
};

void RootMsdc::DpchEngData::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "DpchEngRootMsdcData");
	else
		basefound = checkXPath(docctx, basexpath);

	if (basefound) {
		if (extractStringUclc(docctx, basexpath, "scrJref", "", scrJref)) add(SCRJREF);
		if (feedFEnsSps.readXML(docctx, basexpath, true)) add(FEEDFENSSPS);
	} else {
		feedFEnsSps.clear();
	};
};

File: japimsdc/RootMsdc.java
/**
  * \file RootMsdc.java
  * Java API code for job RootMsdc
  * \author Alexander Wirthmueller
  * \date created: 18 Dec 2018
  * \date modified: 18 Dec 2018
  */

package apimsdc;

import java.util.*;
import org.w3c.dom.*;
import sbecore.*;

public class RootMsdc {
	/**
		* DpchAppLogin (full: DpchAppRootMsdcLogin)
		*/
	public class DpchAppLogin extends DpchAppMsdc {

		public static final int SCRJREF = 1;
		public static final int USERNAME = 2;
		public static final int PASSWORD = 3;
		public static final int M2MNOTREG = 4;
		public static final int CHKSUSPSESS = 5;
		public static final int ALL = 6;

		public DpchAppLogin(
					String scrJref
					, String username
					, String password
					, boolean m2mNotReg
					, boolean chksuspsess
					, Integer[] mask
				) {
			super(VecMsdcVDpch.DPCHAPPROOTMSDCLOGIN, scrJref);

			this.mask = new HashSet<Integer>(Arrays.asList(mask));

			for (Integer i: mask)
				if (i == ALL) {
					this.mask = new HashSet<Integer>(Arrays.asList(SCRJREF, USERNAME, PASSWORD, M2MNOTREG, CHKSUSPSESS));
					break;
				};

			this.username = username;
			this.password = password;
			this.m2mNotReg = m2mNotReg;
			this.chksuspsess = chksuspsess;
		};

		public String username;
		public String password;
		public boolean m2mNotReg;
		public boolean chksuspsess;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(USERNAME)) ss.add("username");
			if (has(PASSWORD)) ss.add("password");
			if (has(M2MNOTREG)) ss.add("m2mNotReg");
			if (has(CHKSUSPSESS)) ss.add("chksuspsess");

			return StrMod.vectorToString(ss, ';');
		};

		public void writeXML(
					Document doc
					, Element sup
				) {
			Element el = doc.createElement("DpchAppRootMsdcLogin");

			if (sup == null) doc.appendChild(el);
			else sup.appendChild(el);

			el.setAttribute("xmlns", "http://www.mpsitech.com");

			if (has(SCRJREF)) Xmlio.writeString(doc, el, "scrJref", scrJref);
			if (has(USERNAME)) Xmlio.writeString(doc, el, "username", username);
			if (has(PASSWORD)) Xmlio.writeString(doc, el, "password", password);
			if (has(M2MNOTREG)) Xmlio.writeBoolean(doc, el, "m2mNotReg", m2mNotReg);
			if (has(CHKSUSPSESS)) Xmlio.writeBoolean(doc, el, "chksuspsess", chksuspsess);
		};

	};

	/**
		* DpchEngData (full: DpchEngRootMsdcData)
		*/
	public class DpchEngData extends DpchEngMsdc {

		public static final int SCRJREF = 1;
		public static final int FEEDFENSSPS = 2;

		public DpchEngData() {
			super(VecMsdcVDpch.DPCHENGROOTMSDCDATA);

			feedFEnsSps = new Feed("FeedFEnsSps");
		};

		public Feed feedFEnsSps;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(FEEDFENSSPS)) ss.add("feedFEnsSps");

			return StrMod.vectorToString(ss, ';');
		};

		public void readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "DpchEngRootMsdcData");

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJref = Xmlio.extractStringUclc(doc, basexpath, "scrJref", "", mask, SCRJREF);
				if (feedFEnsSps.readXML(doc, basexpath, true)) add(FEEDFENSSPS);
			} else {
				scrJref = "";
				feedFEnsSps = new Feed("FeedFEnsSps");
			};
		};

	};

};

